File: typescript/src/server/config/wallet.ts
/**
 * Wallet Configuration Module
 * 
 * Handles loading and validation of wallet configurations for the X402 server.
 * Separates wallet management logic from the main server file for better modularity.
 */

import * as fs from 'fs';
import * as path from 'path';

export interface WalletInfo {
  address: string;
  name: string;
}

/**
 * Load server wallet configuration dynamically from saved wallet data
 * 
 * Reads the server wallet configuration from the JSON file created during setup.
 * This ensures the server uses the correct wallet address for receiving payments.
 * 
 * @returns Server wallet configuration with address and name
 * @throws {Error} When server wallet file is missing or invalid
 */
export function loadServerWallet(): WalletInfo {
  const serverWalletPath = path.join(process.cwd(), 'server-wallet-data.json');
  
  if (!fs.existsSync(serverWalletPath)) {
    console.error('❌ Server wallet not found! Please run: npm run setup');
    console.log('💡 This will create both client and server wallets automatically.');
    process.exit(1);
  }

  try {
    const walletData = JSON.parse(fs.readFileSync(serverWalletPath, 'utf8'));
    
    // Validate wallet data structure
    if (!walletData || typeof walletData !== 'object') {
      throw new Error('Invalid wallet data format');
    }
    
    const serverAddress = walletData.defaultAddress || walletData.addresses?.[0];
    const serverName = walletData.accounts?.[0]?.name || 'Unknown';
    
    if (!serverAddress || !serverAddress.startsWith('0x') || serverAddress.length !== 42) {
      throw new Error('No valid address found in server wallet data');
    }

    console.log(`✅ Server wallet loaded: ${serverAddress}`);
    return { address: serverAddress, name: serverName };
  } catch (error: any) {
    console.error('❌ Failed to load server wallet:', error.message || error);
    console.log('💡 Please run: npm run setup');
    process.exit(1);
  }
}

/**
 * Load client wallet configuration for display purposes
 * 
 * Attempts to load client wallet information for logging and display.
 * This is optional and used only for better user experience in logs.
 * 
 * @returns Client wallet information (may contain placeholder values)
 */
export function loadClientWallet(): WalletInfo {
  const clientWalletPath = path.join(process.cwd(), 'wallet-data.json');
  
  try {
    if (fs.existsSync(clientWalletPath)) {
      const walletData = JSON.parse(fs.readFileSync(clientWalletPath, 'utf8'));
      
      if (walletData && typeof walletData === 'object') {
        const clientAddress = walletData.defaultAddress || walletData.addresses?.[0];
        const clientName = walletData.accounts?.[0]?.name || 'Unknown';
        return { address: clientAddress || 'Not configured', name: clientName };
      }
    }
  } catch {
    console.warn('⚠️ Could not load client wallet info for display');
  }
  
  return { address: 'Not configured', name: 'Unknown' };
} 

File: python/src/server/config/wallet.py
"""
Wallet Configuration Module

Handles loading and validation of server wallet configuration from JSON files.
"""

import json
import os
from pathlib import Path
from typing import Dict, Any, Optional
from dataclasses import dataclass


@dataclass
class WalletConfig:
    """Wallet configuration data structure"""
    id: str
    default_address: str
    addresses: list[str]
    accounts: list[Dict[str, str]]


class WalletConfigLoader:
    """Loads and validates wallet configuration from JSON files"""
    
    def __init__(self, config_path: Optional[str] = None):
        """
        Initialize the wallet config loader
        
        Args:
            config_path: Path to wallet config file (defaults to server-wallet-data.json)
        """
        if config_path is None:
            # Default to server-wallet-data.json in the project root
            project_root = Path(__file__).parent.parent.parent.parent
            config_path = project_root / "server-wallet-data.json"
        
        self.config_path = Path(config_path)
    
    def load(self) -> WalletConfig:
        """
        Load wallet configuration from JSON file
        
        Returns:
            WalletConfig object with validated wallet data
            
        Raises:
            FileNotFoundError: If config file doesn't exist
            ValueError: If config file is invalid or missing required fields
        """
        if not self.config_path.exists():
            raise FileNotFoundError(f"Wallet config file not found: {self.config_path}")
        
        try:
            with open(self.config_path, 'r') as f:
                data = json.load(f)
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid JSON in wallet config file: {e}")
        
        # Validate required fields
        required_fields = ['id', 'defaultAddress', 'addresses', 'accounts']
        for field in required_fields:
            if field not in data:
                raise ValueError(f"Missing required field '{field}' in wallet config")
        
        # Validate default address exists in addresses list
        default_address = data['defaultAddress']
        if default_address not in data['addresses']:
            raise ValueError(f"Default address '{default_address}' not found in addresses list")
        
        return WalletConfig(
            id=data['id'],
            default_address=data['defaultAddress'],
            addresses=data['addresses'],
            accounts=data['accounts']
        )
    
    def get_receiving_address(self) -> str:
        """
        Get the receiving wallet address for X402 payments
        
        Returns:
            The default wallet address to receive payments
            
        Raises:
            FileNotFoundError: If config file doesn't exist
            ValueError: If config file is invalid
        """
        config = self.load()
        return config.default_address


# Global instance for easy access
wallet_config = WalletConfigLoader() 